### PR TITLE
save_analysis: work on HIR tree instead of AST

### PR DESCRIPTION
**Pull request by marmeladema**
In order to reduce the uses of `NodeId`s in the compiler, `save_analysis` crate has been reworked to operate on the HIR tree instead of the AST.

cc #50928

### CODE REVIEW

**Comment by cjgillot** [comment]
It appears you left some dead code commented out. Did you forget to remove it?

**Comment by marmeladema** [comment]
Well originally, i thought tests would fail and that it would take me a while to fix those so i created a draft PR; knowing there would be some cleanups to do. But then tests passed and I converted into a real PR and forgot to remove the dead code.
I'll cleanup it asap :+1:

File: src/librustc_driver/lib.rs [hunk 1]
```diff
@@ -371,8 +369,6 @@ pub fn run_compiler(
                     });
 
                     result
-                    // AST will be dropped *after* the `after_analysis` callback
-                    // (needed by the RLS)
```
Comment (cjgillot) [inline src/librustc_driver/lib.rs:375]: If you remove that comment, can you move the drop of the AST where earlier?

Comment (marmeladema) [reply src/librustc_driver/lib.rs:375]: Will do!

Comment (marmeladema) [reply src/librustc_driver/lib.rs:375]: Done

**Comment by marmeladema** [comment]
I ran the `rls` tests locally and 3 are failing:
* `client_highlight`
* `client_hover`
* `client_rename`

To debug i dumped the analysis using `rustc -Z save-analysis src/tools/rls/tests/fixtures/common/src/main.rs` for both the current nightly and with this change and here what I got:
* Current nightly: https://paste.ee/p/rg5QZ
* With this change: https://paste.ee/p/iMz5W

The only difference is that some id's seem wrong with the current nightly and "fixed' with this change but that does not really explain why the tests are failing.

If we take a closer look at `client_highlight` we can see that the highlight query at `Position { line: 12, character: 27 }` is returning:

```
ranges: [
  DocumentHighlight {
    range: Range { start: Position { line: 10, character: 7 }, end: Position { line: 10, character: 11 } },
    kind: Some(Text)
  },
  DocumentHighlight {
    range: Range { start: Position { line: 12, character: 27 }, end: Position { line: 12, character: 32 } },
    kind: Some(Text)
  },
  DocumentHighlight {
    range: Range { start: Position { line: 15, character: 25 }, end: Position { line: 15, character: 29 } },
    kind: Some(Text)
  }
]
```

Whereas the test expect to find range (no found):

```
Range { start: Position { line: 11, character: 8 }, end: Position { line: 11, character: 13 } }
```

and (found):

```
Range { start: Position { line: 12, character: 27 }, end: Position { line: 12, character: 32 } }
```

I am not sure how to go forward with this or even if the failures are not in fact just fixes?

**Comment by Xanewok** [comment]
That's great, thanks so much for this!

I'll gladly review that tomorrow morning since it's getting late here 😴

**Comment by marmeladema** [comment]
@Xanewok i've updated the pasted links because https://paste.ee seems unavailable right now

File: src/librustc_save_analysis/lib.rs [hunk 2]
```diff
@@ -1109,21 +1070,21 @@ fn id_from_def_id(id: DefId) -> rls_data::Id {
...
-    def_id.map(|id| id_from_def_id(id.to_def_id())).unwrap_or_else(|| {
-        // Create a *fake* `DefId` out of a `NodeId` by subtracting the `NodeId`
-        // out of the maximum u32 value. This will work unless you have *billions*
-        // of definitions in a single crate (very unlikely to actually happen).
-        rls_data::Id { krate: LOCAL_CRATE.as_u32(), index: !id.as_u32() }
-    })
+fn id_from_hir_id(id: hir::HirId, scx: &SaveContext<'_, '_>) -> rls_data::Id {
+    let def_id = scx.tcx.hir().opt_local_def_id(id);
+    def_id
+        .map(|id| id_from_def_id(id.to_def_id()))
```
Comment (Xanewok) [inline src/librustc_save_analysis/lib.rs:1076]: I'm having a hard time grokking the changes here. Why do we fall back to getting def_id from `id.owner`? IIUC the original  code worked around the assumption that node ids are inherently local and may only produce local defs.

Since we have `id: hir::HirId`, can't we directly go from `HirId` to (rls_data-like) `DefId`?

File: src/librustc_save_analysis/dump_visitor.rs [hunk 3]
```diff
@@ -199,41 +199,29 @@ impl<'l, 'tcx> DumpVisitor<'l, 'tcx> {
...
 
-    fn write_sub_paths(&mut self, path: &ast::Path) {
-        for seg in &path.segments {
-            if let Some(data) = self.save_ctxt.get_path_segment_data(seg) {
-                self.dumper.dump_ref(data);
-            }
-        }
-    }
-
     // As write_sub_paths, but does not process the last ident in the path (assuming it
```
[the remark this answers is not in the document]
Comment (marmeladema) [reply src/librustc_save_analysis/dump_visitor.rs:210]: Its in fact needed for the `UseKing::Glob` case

File: src/librustc_save_analysis/lib.rs [hunk 4]
```diff
@@ -666,33 +627,24 @@ impl<'l, 'tcx> SaveContext<'l, 'tcx> {
...
-                    return true;
-                }
-            }
-            false
-        }
-
-        if id == DUMMY_NODE_ID {
-            return None;
+        fn fn_type(seg: &hir::PathSegment<'_>) -> bool {
+            if let Some(ref generic_args) = seg.args { generic_args.parenthesized } else { false }
```
[the remark this answers is not in the document]
Comment (marmeladema) [reply src/librustc_save_analysis/lib.rs]: Fixed!

File: src/librustc_save_analysis/lib.rs [hunk 5]
```diff
@@ -1063,9 +1023,10 @@ pub fn process_crate<'l, 'tcx, H: SaveHandler>(
 
         let mut visitor = DumpVisitor::new(save_ctxt);
 
-        visitor.dump_crate_info(cratename, krate);
+        visitor.dump_crate_info(cratename, tcx.hir().krate());
         visitor.dump_compilation_options(input, cratename);
-        visit::walk_crate(&mut visitor, krate);
+        // intravisit::walk_crate(&mut visitor, krate);
```
[the remark this answers is not in the document]
Comment (marmeladema) [reply src/librustc_save_analysis/lib.rs]: Done

File: src/librustc_save_analysis/dump_visitor.rs [hunk 6]
```diff
@@ -1076,177 +1047,61 @@ impl<'l, 'tcx> DumpVisitor<'l, 'tcx> {
...
-
-                // Otherwise it's a span with wrong macro expansion info, which
-                // we don't want to track anyway, since it's probably macro-internal `use`
-                if let Some(sub_span) =
-                    self.span.sub_span_of_token(use_tree.span, token::BinOp(token::Star))
-                {
-                    if !self.span.filter_generated(use_tree.span) {
-                        let span = self.span_from_span(sub_span);
-
-                        self.dumper.import(
```
[the remark this answers is not in the document]
Comment (marmeladema) [reply src/librustc_save_analysis/dump_visitor.rs:1197]: I re-added it, i hope its fine now :crossed_fingers:

File: src/librustc_save_analysis/lib.rs (hunk 2 above)
Comment (marmeladema) [reply src/librustc_save_analysis/lib.rs:1075]: I am not sure what exactly rls needs. Internally the compiler use `(opt_)local_def_id` to lookup the `LocalDefId` associated with some `HirId` because not all `LocalDefId` are `HirId` owners and especially because the real owner can be different from `hir_id.owner`. The fallback case (using `hir_id.owner`) should actually almost never happen in practice its more like a safe guard for very weird bugs in the compiler where a `HirId` does not actually have an associated `LocalDefId` at all.

Comment (Xanewok) [reply src/librustc_save_analysis/lib.rs:1075]: > I am not sure what exactly rls needs.

To be 100% frank I can't recall the reasons behind why we're creating `rls_data::Id`s like that :sweat_smile: 

> (...) its more like a safe guard for very weird bugs in the compiler where a `HirId` does not actually have an associated `LocalDefId` at all.

Could we document that somewhere here?

>  (...) because the real owner can be different from hir_id.owner (...)

Could you give an example, so I could better understand the situation?

Otherwise, this looks good! I'd personally very much appreciate appreciate a `match` with some specific comments like why there can be ever a `None` case or why `id.owner` is a good fallback :slightly_smiling_face: It's somewhat hard to follow along, otherwise

File: src/librustc_save_analysis/dump_visitor.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Xanewok) [reply src/librustc_save_analysis/dump_visitor.rs:1197]: Thanks!

File: src/librustc_save_analysis/lib.rs (hunk 2 above)
Comment (Xanewok) [reply src/librustc_save_analysis/lib.rs:1075]: Okay, so this sheds some more light as to why we use two-sided index. If we take a look at the save-analysis generated with this patch, we can see:
```
    {
      "attributes": [],
      "children": [],
      "decl_id": null,
      "docs": "",
      "id": {
        "index": 5,
        "krate": 0
      },
      "kind": "Local",
      "name": "world",
      "parent": null,
      "qualname": "world$18",
      "sig": null,
      "span": {
        "byte_end": 147,
        "byte_start": 142,
        "column_end": 14,
        "column_start": 9,
        "file_name": "src/tools/rls/tests/fixtures/common/src/main.rs",
        "line_end": 12,
        "line_start": 12
      },
      "value": "&str"
    },
    {
      "attributes": [],
      "children": [],
      "decl_id": null,
      "docs": "",
      "id": {
        "index": 5,
        "krate": 0
      },
      "kind": "Local",
      "name": "bar2",
      "parent": null,
      "qualname": "bar2$21",
      "sig": null,
      "span": {
        "byte_end": 207,
        "byte_start": 203,
        "column_end": 13,
        "column_start": 9,
        "file_name": "src/tools/rls/tests/fixtures/common/src/main.rs",
        "line_end": 15,
        "line_start": 15
      },
      "value": "Bar"
```

That's two "definitions", which have the same `def_id`. Informally, these are definitions in the name resolution sense (which don't have a separate DefId IIUC) and not item-like sense (which should have DefId). For the former, save-analysis used a negated index, so it created a *fake* def ids for those.

Because of this, I think we should retain the old behaviour for now (synthesizing `rls_data::Id using negative indexing of `HirId`/`NodeId` as a numeric value) in case we don't have a `LocalDefId` for a given `HirId`.

Comment (marmeladema) [reply src/librustc_save_analysis/lib.rs:1075]: Ok that makes perfect sense! I'll into that.
On a side note, i wondered if, before merging this, we should use all the fixture test files as test in rustc itself and generate the dumped analysis for all of those and checked for regressions. Iirc miri does a similar things. This way we would have a better confidence about the impact of the changes for this work.

Comment (marmeladema) [reply src/librustc_save_analysis/lib.rs:1075]: About `HirId` being owned by a different `LocalDefId` than `hir_id.owner`, you can take a look at my comment here: https://github.com/rust-lang/rust/issues/50928#issuecomment-635232363
TL;DR is closures behave like that

Comment (Xanewok) [reply src/librustc_save_analysis/lib.rs:1075]: That's a great idea in general. Because this is slowly becoming legacy and unloved code, there was an idea floating around that we should remove save_analysis at some point altogether (mb that's not the case anymore with this HIR rewrite?) - because of this, I'm not sure if this is worth the investment.

Addditionally, the IDs are not guaranteed to be stable (rn based on node ids as you've seen) so I'd be afraid this might induce more test update churn etc. If we can easily come up with a solution that verifies the relations and aux data defined in these files rather than concrete values, then I'd be up for it!

I wouldn't want to block this PR on it, though.

Comment (marmeladema) [reply src/librustc_save_analysis/lib.rs]: I fixed the id generation and indeed it fixes `rls` tests

**Comment by marmeladema** [comment]
@Xanewok anything else i should take care of?

File: src/librustc_save_analysis/lib.rs (hunk 2 above)
Comment (Xanewok) [reply src/librustc_save_analysis/lib.rs]: Looks great, thanks!

**Comment by Xanewok** [comment]
Thank you for your work and for your patience! I'd say this looks good to go :)

Since this has a potential to cause some issues (it's a considerable rewrite!), I'd sleep better if we waited for the beta cut (happening around now) and then land this. Does that sound good?

**Comment by marmeladema** [comment]
@Xanewok beta has landed i believe. Once this is merged it will break RLS from compiling because the prototype of `process_crate` changed. Is https://github.com/rust-lang/rls the right repo to provide a fix?

**Comment by Xanewok** [comment]
Yep, that’s the one!

Since you’re testing the RLS already and have a fix, could you submit a PR
there and we coordinate the push together?

I can push the relevant RLS update to a separate branch, we’ll include the
submodule update in this PR to that branch and after that I’ll merge the
RLS branch to its master.

This way we can avoid the breakage altogether 👍

**Comment by marmeladema** [comment]
@Xanewok here is the PR for `rls`: https://github.com/rust-lang/rls/pull/1677

**Comment by Xanewok** [comment]
Merged into a separate branch: https://github.com/rust-lang/rls/commit/8d7a7167c15b9154755588c39b22b2336c89ca68

Could you include the submodule bump in this PR?

All you need to do is to: `git checkout 8d7a7167c15b9154755588c39b22b2336c89ca68` in `src/tools/rls` and then `git add src/tools/rls` from the root followed by a commit :smiley:

**Comment by marmeladema** [comment]
@Xanewok done :+1:

**Comment by Xanewok** [comment]
Awesome, thank you!

@bors r+

**Comment by Xanewok** [comment]
Let's bump the prio since this updates the submodule as well:

@bors p=1

File: src/librustc_save_analysis/sig.rs [hunk 7]
```diff
@@ -145,49 +153,52 @@ fn text_sig(text: String) -> Signature {
     Signature { text, defs: vec![], refs: vec![] }
 }
 
-fn push_extern(text: &mut String, ext: Extern) {
+/*fn push_extern(text: &mut String, ext: Extern) {
```
Comment (cjgillot) [inline src/librustc_save_analysis/sig.rs:156]: You forgot to remove this dead code.

**Comment by cjgillot** [comment]
There are still a few FIXME in the code. Do you plan to address them before merging?

File: src/librustc_save_analysis/sig.rs (hunk 7 above)
Comment (Xanewok) [reply src/librustc_save_analysis/sig.rs:156]: Nice catch! I scanned the code last time and somehow missed that and thought it's resolved, let's wait with this for now

Comment (Xanewok) [reply src/librustc_save_analysis/sig.rs:156]: (hopefully bors works in the review comments)

@bors r-

Comment (marmeladema) [reply src/librustc_save_analysis/sig.rs:156]: Oh sorry fixing it now!

**Comment by marmeladema** [comment]
Should be good now hopefully.

**Comment by marmeladema** [comment]
@Xanewok failure looks spurious. Just rebased for the sake of it.

**Comment by jonas-schievink** [comment]
Looks like https://github.com/rust-lang/rust/issues/71099

@bors r=Xanewok

File: src/librustc_driver/lib.rs [hunk 8]
```diff
@@ -346,20 +346,22 @@ pub fn run_compiler(
 
             queries.global_ctxt()?;
 
+            // Drop AST after creating GlobalCtxt to free memory
+            let _timer = sess.prof.generic_activity("drop_ast");
+            mem::drop(queries.expansion()?.take());
```
Comment (eddyb) [inline src/librustc_driver/lib.rs:351]: This needs a block around it, the `_timer` is supposed to be dropped after the next statement, in order for `drop_ast` to track only that, and not the rest of this large block. (it's not a big deal, but I was baffled as to why `drop_ast` had so much variance, or any children at all, in the profiling data)

Comment (marmeladema) [reply src/librustc_driver/lib.rs:351]: Sorry about that! Should be fixed by https://github.com/rust-lang/rust/pull/75575
